SCClassLibrary/External/Ableton/LinkClock.sc

```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
+					);
+				};
+			}, '/LinkClock/changeMeter');
+		} {
+			meterChangeResp.free;
+		};
+	}
+
+	setMeterAtBeat { |newBeatsPerBar, beats|
+		if(syncMeter) {
+			NetAddr.broadcastFlag = true;
+			(57120..57127).do { |port|
+				NetAddr("255.255.255.255", port).sendMsg(
+					'/LinkClock/changeMeter', id, newBeatsPerBar, this.beats, beats
+				);
+			};
+		};
+		this.prSetMeterAtBeat(newBeatsPerBar, beats);
+	}
+
+	queryMeter { |action, timeout = 0.2|
+		var replies = IdentityDictionary.new,
+		resp;
+		if(this.numPeers > 0) {
+			resp = OSCFunc({ |msg, time, addr|
+				if(msg[1] != id) {
+					replies.put(msg[1], (
+						id: msg[1],
+						queriedAtBeat: msg[2],
+						syncMeter: msg[3].asBoolean,
+						beats: msg[4],
+						beatsPerBar: msg[5],
+						baseBarBeat: msg[6],
+						beatInBar: msg[7]
+					));
+				};
+			}, '/LinkClock/meterReply');
+			{
+				NetAddr.broadcastFlag = true;
+				// maybe some peers had to take a different port
+				(57120 .. 57127).do { |port|
+					NetAddr("255.255.255.255", port).sendMsg('/LinkClock/queryMeter', this.beats);
+				};
+				(timeout * this.tempo).wait;  // seconds, not beats
+				resp.free;
+				action.value(replies.values);
+			}.fork(this);
+		} {
+			// no peers, don't bother asking
+			// sched is needed to make sure Conditions waiting for this method still work
+			this.sched(0, { action.value(List.new) });
+		};
+	}
+
+	adjustMeterBase { |localBeats, remoteBeats, round = 1|
+		baseBarBeat = baseBarBeat + ((localBeats - remoteBeats) % beatsPerBar).round(round);
+	}
+
+	// if there are peers with syncMeter = true, align my clock to their common barline
+	// ignore peers with syncMeter = false
+	resyncMeter { |round, verbose = true|
+		var replies, cond = Condition.new, bpbs, baseBeats, denom;
+		var newBeatsPerBar, newBase;
+		fork {
+			this.queryMeter { |val|
+				replies = val;
+				cond.unhang;
+			};
+			cond.hang;
+			replies = replies.select { |reply| reply[\syncMeter] };
+			if(replies.size > 0) {
+				// verify beatsPerBar, and myBeats - beatInBar, are common across peers
+				// (if the network is glitchy, we may have replies from different trials.
+				// but the difference between query time and remote beatInBar should be
+				// consistent.)
+				// also calculate baseBarBeat common denominator
+				bpbs = Set.new;
+				baseBeats = Set.new;
+				denom = 1;
+				replies.do { |reply, i|
+					denom = denom lcm: reply[\baseBarBeat].asFraction[1];
+					bpbs.add(reply[\beatsPerBar]);
+					baseBeats.add(reply[\queriedAtBeat] - reply[\beatInBar]);
+				};
+				if(round.isNil) { round = denom.reciprocal };
+				// Now it gets tricky. We need to reduce baseBeats based on beatsPerBar.
+				// All 'syncMeter' peers should have the same beatsPerBar (size == 1).
+				// But, if something failed, maybe there are more.
+				// So we have to do a two-step check.
+				if(bpbs.size == 1) {
+					// 'choose' = easy way to get one item from an unordered collection
+					newBeatsPerBar = bpbs.choose;
```

Comment:
`pop` is better, i think. `choose` randomly probes the backing array until a valid item is found, while `pop` will linearly scan. since items are inserted linearly into the backing array, `pop` will return after checking the first index while `choose` may not

Reply:
Good catch, will do.